crates/matrix-sdk/src/event_cache/deduplicator.rs

```diff
@@ -60,16 +61,47 @@ impl Deduplicator {
 
     /// Find duplicates in the given collection of events, and return both
     /// valid events (those with an event id) as well as the event ids of
-    /// duplicate events.
+    /// duplicate events along with their position.
     pub async fn filter_duplicate_events(
         &self,
-        events: Vec<Event>,
+        mut events: Vec<Event>,
         room_events: &RoomEvents,
-    ) -> Result<(Vec<Event>, Vec<OwnedEventId>), EventCacheError> {
-        match self {
-            Deduplicator::InMemory(dedup) => Ok(dedup.filter_duplicate_events(events, room_events)),
-            Deduplicator::PersistentStore(dedup) => dedup.filter_duplicate_events(events).await,
+    ) -> Result<DeduplicationOutcome, EventCacheError> {
+        // Remove all events with no ID, or that is duplicated inside `events`, i.e.
+        // `events` contains duplicated events in itself, e.g. `[$e0, $e1, $e0]`, here
+        // `$e0` is duplicated in within `events`.
+        {
+            let mut event_ids = BTreeSet::new();
+
+            events.retain(|event| {
+                let Some(event_id) = event.event_id() else {
+                    // No event ID? Bye bye.
```

Comment:
Under what circumstances would an event passed into here not have an event it? 🤔